tests: Add fuzzing harness for CAddrMan

practicalswift opened the pull request:
Add fuzzing harness for `CAddrMan`.

~~Fill some fuzzing coverage gaps for functions in `addrdb.h`, `merkleblock.h` and `outputtype.h`.~~

See [`doc/fuzzing.md`](https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md) for information on how to fuzz Bitcoin Core. Don't forget to contribute any coverage increasing inputs you find to the [Bitcoin Core fuzzing corpus repo](https://github.com/bitcoin-core/qa-assets).

Happy fuzzing :)

practicalswift committed:
tests: Add fuzzing harness for CAddrMan

MarcoFalke commented on src/test/fuzz/kitchen_sink.cpp:
```diff
@@ -2,15 +2,27 @@
+void initialize()
+{
+    SelectParams(CBaseChainParams::REGTEST);
+    StartMutedLogging();
```
Why can't this be a testing setup with logging disabled, like in the other tests?

practicalswift replied:
Unfortunately `TestingSetup` introduces non-determinism which makes coverage vary between runs, see https://github.com/bitcoin/bitcoin/pull/19067#discussion_r429986672 :)

MarcoFalke replied:
What about BasicTestgingSetup then?

practicalswift replied:
I think it is the same problem there: I get more setup than I need and a subset of that setup work introduces coverage non-determinism :)

I prefer opt-ing in to exactly the setup I need: it avoids the non-determinism problem and I also think that makes it easier to reason about the dependencies of each fuzzing harness.

MarcoFalke replied:
If non-determinism is an issue for fuzz tests, it is also an issue for unit tests etc. Instead of solving it fresh in every single test case, I'd rather fix it once and for all.

Using hacks that don't clean up after themselves, depend on global state and leave global state dirty is asking for mishaps. Currently the fuzz datadir is set to the default location of the datadir. I don't think I need to repeat why this is bad.

If the basic testing setup is still too bloated, then please add one that is stripped down, so that unit and bench tests can use it as well.

practicalswift replied:
I was wrong about `BasicTestingSetup`: it seems like only `TestingSetup` introduces coverage non-determinism.

This shows the issue with `TestingSetup`:

```
# With TestingSetup:
$ src/test/fuzz/kitchen_sink corpus/
…
#20425  NEW    cov: 3867 ft: 12141 corp: 248/49Kb lim: 607 exec/s: 464 rss: 357Mb L: 605/607 MS: 5 InsertRepeatedBytes-CopyPart-InsertByte-ChangeBinInt-CopyPart-
        NEW_FUNC[1/68]: 0x56533ed75c60 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::c_str() const /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/basic_string.h:2291
        NEW_FUNC[2/68]: 0x56533f03dfc0 in std::function<void ()>::function(std::function<void ()> const&) /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/std_function.h:655
#20441  NEW    cov: 4384 ft: 12690 corp: 249/49Kb lim: 607 exec/s: 454 rss: 358Mb L: 326/607 MS: 1 CopyPart-
…
#28628        NEW    cov: 4384 ft: 12731 corp: 258/54Kb lim: 652 exec/s: 454 rss: 452Mb L: 652/652 MS: 1 CrossOver-
==6668== libFuzzer: run interrupted; exiting

$ src/test/fuzz/kitchen_sink corpus/
#423    INITED cov: 3867 ft: 12182 corp: 245/48Kb exec/s: 423 rss: 148Mb
```

Notice the non-deterministic coverage increase from 3867 to 4384 which drops back to 3867 when the fuzzer is restarted.

I'll go with `BasicTestingSetup` :)

practicalswift replied:
Coverage stability with `BasicTestingSetup` vs `TestingSetup`:

Fuzzer with an empty `test_one_input` function, and an `initialize` function containing `static const BasicTestingSetup basic_testing_setup`:

```
INFO: A corpus is not provided, starting from an empty corpus
#2      INITED cov: 58 ft: 59 corp: 1/1b exec/s: 0 rss: 125Mb
#65536  pulse  cov: 58 ft: 59 corp: 1/1b lim: 652 exec/s: 21845 rss: 145Mb
#131072 pulse  cov: 58 ft: 59 corp: 1/1b lim: 1300 exec/s: 21845 rss: 188Mb
#262144 pulse  cov: 58 ft: 59 corp: 1/1b lim: 2611 exec/s: 21845 rss: 339Mb
#524288 pulse  cov: 58 ft: 59 corp: 1/1b lim: 4096 exec/s: 20164 rss: 715Mb
#1048576        pulse  cov: 58 ft: 59 corp: 1/1b lim: 4096 exec/s: 20560 rss: 716Mb
#2097152        pulse  cov: 58 ft: 59 corp: 1/1b lim: 4096 exec/s: 20560 rss: 716Mb
```

Fuzzer with an empty `test_one_input` function, and an `initialize` function containing `static const TestingSetup testing_setup`:

```
INFO: A corpus is not provided, starting from an empty corpus
#2      INITED cov: 58 ft: 59 corp: 1/1b exec/s: 0 rss: 134Mb
#65536  pulse  cov: 58 ft: 59 corp: 1/1b lim: 652 exec/s: 21845 rss: 155Mb
#131072 pulse  cov: 58 ft: 59 corp: 1/1b lim: 1300 exec/s: 21845 rss: 198Mb
#262144 pulse  cov: 58 ft: 59 corp: 1/1b lim: 2611 exec/s: 21845 rss: 348Mb
#524288 pulse  cov: 58 ft: 59 corp: 1/1b lim: 4096 exec/s: 20164 rss: 715Mb
        NEW_FUNC[1/85]: 0x55e526ad96c0 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::allocator<char> >(char const*, std::allocator<char> const&) /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/basic_string.h:516
        NEW_FUNC[2/85]: 0x55e526adc5a0 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_local_data() /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/basic_string.h:180
#910854 NEW    cov: 663 ft: 143 corp: 2/3b lim: 4096 exec/s: 20241 rss: 717Mb L: 2/2 MS: 2 InsertByte-CopyPart-
#1048576        pulse  cov: 663 ft: 143 corp: 2/3b lim: 4096 exec/s: 20560 rss: 717Mb
        NEW_FUNC[1/10]: 0x55e526ad95c0 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::c_str() const /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/basic_string.h:2291
        NEW_FUNC[2/10]: 0x55e526adc140 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose() /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/basic_string.h:219
#1830962        NEW    cov: 711 ft: 335 corp: 3/4099b lim: 4096 exec/s: 20572 rss: 717Mb L: 4096/4096 MS: 3 ChangeBinInt-CrossOver-ChangeASCIIInt-
#2097152        pulse  cov: 711 ft: 335 corp: 3/4099b lim: 4096 exec/s: 20360 rss: 717Mb
```

practicalswift commented:
Anything left to do here? :)  The changes are limited to `src/test/fuzz/` and should hopefully be trivial to review.

Crypt-iQ commented:
I cannot build this branch with either `clang/clang++` or `afl-gcc/afl-g++`:

clang version: `6.0.0-1ubuntu2 (tags/RELEASE_600/final)
afl-gcc/afl-g++ built from master
```
Making all in src
  2 make[1]: Entering directory '/root/bitcoin/src'
  3 make[2]: Entering directory '/root/bitcoin/src'
  4   CXX      test/fuzz/test_fuzz_addition_overflow-addition_overflow.o
  5 In file included from test/fuzz/addition_overflow.cpp:7:
  6 ./test/fuzz/util.h:62:18: error: no template named 'optional' in namespace 'std'; did you mean 'Optional'?
  7 NODISCARD inline std::optional<T> ConsumeDeserializable(FuzzedDataProvider& fuzzed_data_provider, const si    ze_t max_length = 4096) noexcept
  8                  ^~~~~~~~~~~~~
  9                  Optional
 10 ./optional.h:14:1: note: 'Optional' declared here
 11 using Optional = boost::optional<T>;
 12 ^
 13 In file included from test/fuzz/addition_overflow.cpp:7:
 14 ./test/fuzz/util.h:70:16: error: no member named 'nullopt' in namespace 'std'; did you mean simply 'nullop    t'?
 15         return std::nullopt;
 16                ^~~~~~~~~~~~
 17                nullopt
 18 ./optional.h:24:14: note: 'nullopt' declared here
 19 static auto& nullopt = boost::none;
 20              ^
 21 2 errors generated.
 22 Makefile:14081: recipe for target 'test/fuzz/test_fuzz_addition_overflow-addition_overflow.o' failed
 23 make[2]: *** [test/fuzz/test_fuzz_addition_overflow-addition_overflow.o] Error 1
 24 make[2]: Leaving directory '/root/bitcoin/src'
 25 Makefile:17268: recipe for target 'all-recursive' failed
 26 make[1]: *** [all-recursive] Error 1
 27 make[1]: Leaving directory '/root/bitcoin/src'
 28 Makefile:785: recipe for target 'all-recursive' failed
 29 make: *** [all-recursive] Error 1
```

MarcoFalke commented:
@Crypt-iQ So to clarify, you can build current `master`, but not this pull request merged into current `master`?

Crypt-iQ commented:
@MarcoFalke I can build the current `master` branch, but not this branch as it is. I checkout this branch via:

```
git fetch origin pull/19065/head:addrmanfuzz
```

practicalswift commented:
@Crypt-iQ Oh, that's weird! Thanks for reporting. I've now rebased this PR on current `master`. Would you mind testing again? :)

Crypt-iQ commented:
@practicalswift I think you may have forgotten to push it up :) And will test once it's up

practicalswift commented:
@Crypt-iQ Oh, thanks for letting me know! Now pushed for real :)

Crypt-iQ commented:
@practicalswift Able to build this branch with afl and with libfuzzer (clang-10)

Crypt-iQ commented on src/test/fuzz/merkleblock.cpp:
```diff
@@ -16,12 +16,49 @@
+        std::set<uint256> txids;
+        if (opt_block && !opt_block->vtx.empty()) {
+            if (fuzzed_data_provider.ConsumeBool()) {
+                merkle_block = CMerkleBlock{*opt_block, bloom_filter};
```
Could a case be made to not only have the default, match-all filter?

practicalswift replied:
Fixed! I'm now covering all three:
* `CMerkleBlock{*opt_block, bloom_filter}`
* `CMerkleBlock{*opt_block, txids}`
* `CMerkleBlock{}` (added)

Crypt-iQ reviewed (commented):
Still fuzzing `addrman` to hit the collision code, will report back. Fuzzed `merkleblock` & `kitchensink` for ~32 hours each with libfuzzer and `--with-sanitizers=address,fuzzer,undefined`: https://crypt-iq.github.io/merkleblock_and_kitchensink_cov/src/index.html

Only two minor comments.

practicalswift commented on src/test/fuzz/addrman.cpp:
```diff
@@ -0,0 +1,113 @@
+            (void)addr_man.size();
+
+            CDataStream data_stream(SER_NETWORK, PROTOCOL_VERSION);
+            data_stream << addr_man;
```
> Is the other way `>>` skipped because it's covered in `deserialize.cpp`?

Yes :)

> Also since `m_asmap` isn't specified there is a single line that the fuzzer misses:

Thanks! Now fixed.

practicalswift commented:
@Crypt-iQ Thanks a lot for reviewing! Feedback addressed. Would you mind re-reviewing? :)

practicalswift commented:
@MarcoFalke Would you mind reviewing? FWIW this PR touches only `src/test/fuzz/` :)

Crypt-iQ commented:
The `CAddrMan` harness looks good to me. 
Ubuntu 18:
- `./configure --enable-fuzz --with-sanitizers=address,undefined,integer,fuzzer` reports implicit-signed-integer-truncation
- valgrind reports no errors

macOS v10.15.4:
- `./configure --enable-fuzz --with-sanitizers=address,fuzzer --disable-asm` reports no errors
- `./configure --enable-fuzz --with-sanitizers=undefined,integer,fuzzer` reports implicit-signed-integer-truncation

<details>
<summary>implicit-signed-integer-truncation</summary>

```
addrman.cpp:524:22: runtime error: implicit conversion from type 'int64_t' (aka 'long long') of value 60031562745 (64-bit, signed) to type 'uint32_t' (aka 'unsigned int') changed the value to 4196987897 (32-bit, unsigned)
    #0 0x103319c69 in CAddrMan::Connected_(CService const&, long long) addrman.cpp:524
    #1 0x1032a0a1b in CAddrMan::Connected(CService const&, long long) addrman.h:652
    #2 0x10329c621 in test_one_input(std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > const&) addrman.cpp:89
    #3 0x1042c8ff8 in LLVMFuzzerTestOneInput fuzz.cpp:45
    #4 0x1044cb970 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) FuzzerLoop.cpp:556
    #5 0x1044cb0b5 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) FuzzerLoop.cpp:470
    #6 0x1044cd757 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__1::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) FuzzerLoop.cpp:765
    #7 0x1044cdab9 in fuzzer::Fuzzer::Loop(std::__1::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) FuzzerLoop.cpp:792
    #8 0x1044baf7d in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) FuzzerDriver.cpp:829
    #9 0x1044e6c52 in main FuzzerMain.cpp:19
    #10 0x7fff73778cc8 in start+0x0 (libdyld.dylib:x86_64+0x1acc8)

SUMMARY: UndefinedBehaviorSanitizer: implicit-signed-integer-truncation addrman.cpp:524:22 in
```
</details>

Still couldn't hit the collision code, so may need to run for longer. Will review and run the changes to the other harnesses.

Crypt-iQ commented on src/test/fuzz/addrman.cpp:
```diff
@@ -0,0 +1,114 @@
+            const std::optional<CAddress> opt_address = ConsumeDeserializable<CAddress>(fuzzed_data_provider);
+            const std::optional<CNetAddr> opt_net_addr = ConsumeDeserializable<CNetAddr>(fuzzed_data_provider);
+            if (opt_address && opt_net_addr) {
+                addr_man.Add(*opt_address, *opt_net_addr, fuzzed_data_provider.ConsumeIntegralInRange<int64_t>(0, 100000000));
```
How was 100000000 chosen?

practicalswift replied:
100 000 000  was chosen to avoid this signed integer overflow:

```
addrman.cpp:306:67: runtime error: signed integer overflow: 100000000 - -9223372036854775808 cannot be represented in type 'long'
    #0 0x5594ad96b160 in CAddrMan::Add_(CAddress const&, CNetAddr const&, long) src/addrman.cpp:306:67
    #1 0x5594ad90d6ce in CAddrMan::Add(CAddress const&, CNetAddr const&, long) src/./addrman.h:556:17
    #2 0x5594ad909041 in test_one_input(std::vector<unsigned char, std::allocator<unsigned char> > const&) src/test/fuzz/addrman.cpp:53:26
```

Crypt-iQ commented:
The merkleblock and kitchensink harnesses look good to me. Fuzzers didn't report any errors nor with valgrind.

So the reason the `addrman` fuzzer wasn't able to hit the collision code *consistently* is because of `insecure_rand` which changes across runs. Another source of test randomness is the `GetAdjustedTime` function.

Can also see it when running `make cov_fuzz` twice as the `cov` and `ft` numbers differ with libfuzzer output.
First run:
```
Run addrman with args ['/Users/nsa/bitcoin_cov/src/test/fuzz/addrman', '-runs=1', 'qa-assets/fuzz_seed_corpus/addrman']INFO: Seed: 1286454130
INFO: Loaded 1 modules   (264114 inline 8-bit counters): 264114 [0x1114a3ee8, 0x1114e469a), 
INFO: Loaded 1 PC tables (264114 PCs): 264114 [0x1114e46a0,0x1118ec1c0), 
INFO:     5557 files found in qa-assets/fuzz_seed_corpus/addrman
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4151 bytes
INFO: seed corpus: files: 5557 min: 1b max: 4151b total: 3492077b rss: 44Mb
#1024	pulse  cov: 3008 ft: 8297 corp: 326/14217b exec/s: 341 rss: 45Mb
#2048	pulse  cov: 3127 ft: 10801 corp: 530/40Kb exec/s: 89 rss: 45Mb
#4096	pulse  cov: 3160 ft: 13578 corp: 830/143Kb exec/s: 42 rss: 45Mb
#5558	INITED cov: 3172 ft: 15459 corp: 1160/728Kb exec/s: 29 rss: 46Mb
#5558	DONE   cov: 3172 ft: 15459 corp: 1160/728Kb lim: 4147 exec/s: 29 rss: 46Mb
Done 5558 runs in 191 second(s)
```

Second run:
```
Run addrman with args ['/Users/nsa/bitcoin_cov/src/test/fuzz/addrman', '-runs=1', 'qa-assets/fuzz_seed_corpus/addrman']INFO: Seed: 1987669807
INFO: Loaded 1 modules   (264114 inline 8-bit counters): 264114 [0x104dd7ee8, 0x104e1869a), 
INFO: Loaded 1 PC tables (264114 PCs): 264114 [0x104e186a0,0x1052201c0), 
INFO:     5557 files found in qa-assets/fuzz_seed_corpus/addrman
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4151 bytes
INFO: seed corpus: files: 5557 min: 1b max: 4151b total: 3492077b rss: 44Mb
#1024	pulse  cov: 3008 ft: 8316 corp: 330/14641b exec/s: 204 rss: 45Mb
#2048	pulse  cov: 3117 ft: 10818 corp: 522/39Kb exec/s: 49 rss: 45Mb
#4096	pulse  cov: 3159 ft: 13573 corp: 838/148Kb exec/s: 27 rss: 45Mb
Slowest unit: 12 s:
artifact_prefix='./'; Test unit written to ./slow-unit-8450b39fa9bc5a9408007860514c53b0b6bf2723
#5558	INITED cov: 3181 ft: 15428 corp: 1159/714Kb exec/s: 20 rss: 46Mb
#5558	DONE   cov: 3181 ft: 15428 corp: 1159/714Kb lim: 4147 exec/s: 20 rss: 46Mb
Done 5558 runs in 267 second(s)
```

practicalswift commented:
@Crypt-iQ I'm afraid we'll have to live with some degree of non-determinism. With that said I'm open to all suggestions on how to reduce the non-determinism in testing, but I guess that is out of scope for this PR :)

Crypt-iQ commented on src/test/fuzz/addrman.cpp:
```diff
@@ -0,0 +1,114 @@
+    (void)addr_man.size();
+    if (fuzzed_data_provider.ConsumeBool()) {
+        while (fuzzed_data_provider.ConsumeBool()) {
+            addr_man.m_asmap.push_back(fuzzed_data_provider.ConsumeBool());
```
Should this be at the beginning so that addrman GetGroup calls use a populated asmap?

practicalswift replied:
Thanks! Fixed :)

MarcoFalke commented:
> So the reason the addrman fuzzer wasn't able to hit the collision code consistently is because of insecure_rand which changes across runs. Another source of test randomness is the GetAdjustedTime function.

I presume `GetAdjustedTime` is non-deterministic because you it asks the system for the time. With `SetMockTime` you can make it return the time you (or the fuzz engine) set.

`insecure_rand` could be made deterministic by accepting a bool parameter to the addrman constructor and passing it on to `insecure_rand`.

Emzy commented:
tACK
Tested and worked:
src/test/fuzz/addrman
src/test/fuzz/kitchen_sink
src/test/fuzz/merkleblock

fjahr commented:
tested ACK 4bd95f42e2d18bdf86c05f4b9b5e31fcf0e3ba1d

Reviewed code and ran the affected fuzz tests for a while.

practicalswift commented:
Ready for merge? :)

PR touches code only in `src/test/fuzz/` and has two Tested ACK:s.

Crypt-iQ commented:
@practicalswift Just had that one comment remaining about `GetGroup`. I tried with @MarcoFalke suggested patch of `insecure_rand` being deterministic and after about a week the fuzzer still did not hit the collision code according to lcov. Granted, I am not throwing very many cores at it (2), and some input could probably be engineered to trigger it - just haven't had the time. As a newcomer to the codebase, I don't know what is considered out of scope PR-wise, but test determinism is always nice-to-have.

practicalswift commented:
@Crypt-iQ Thanks for clarifying! I've now moved the AS-map code to allow for use via `GetGroup`. I'm now also mocking the time, but I skipped the suggested `insecure_rand` since I don't want to change any code outside of `src/test/fuzz/` in this PR. That can be taken in a follow-up if deemed necessary.

I think this PR should be ready for final review now. Many thanks for your patience: may I ask you to review for the n:th time? :)

MarcoFalke commented:
Concept ACK. Will review

MarcoFalke commented on src/test/fuzz/addrman.cpp:
```diff
@@ -0,0 +1,120 @@
+    if (fuzzed_data_provider.ConsumeBool()) {
+        while (fuzzed_data_provider.ConsumeBool()) {
+            addr_man.m_asmap.push_back(fuzzed_data_provider.ConsumeBool());
+        }
```
Encoding one bit as two bytes in the seed sounds really inefficient

practicalswift replied:
Now optimized by using `BytesToBits`.

MarcoFalke commented on src/test/fuzz/merkleblock.cpp:
```diff
@@ -16,12 +16,49 @@
+
+    {
+        CMerkleBlock merkle_block;
+        const std::optional<CBlock> opt_block = ConsumeDeserializable<CBlock>(fuzzed_data_provider);
```
It doesn't look like there is any state shared between the two fuzz targets in this file. Often the fuzz engine can only generate random noise and I write initial seeds myself. With a format that has several unrelated seeds concatenated together, this becomes harder.

I guess the options are:
* Split into two fuzz scripts (requires one more binary)
* Use a switch-case to select a fuzz target within the script (requires one more byte at the end of each fuzz input)

practicalswift replied:
Thanks for reviewing. Now skipping all changes that are unrelated to CAddrMan to keep this PR focused.

MarcoFalke commented on src/test/fuzz/kitchen_sink.cpp:
```diff
@@ -23,4 +34,21 @@ void test_one_input(const std::vector<uint8_t>& buffer)
+    (void)ParseOutputType(fuzzed_data_provider.ConsumeRandomLengthString(64), output_type_parsed);
+
+    CAddrMan addr_man;
+    CDataStream data_stream_peers = ConsumeDataStream(fuzzed_data_provider);
```
Same here. Creating an addrman stream seed by hand requires a random length string first.

practicalswift replied:
Thanks for reviewing. Now skipping all changes that are unrelated to CAddrMan to keep this PR focused.

practicalswift commented:
@MarcoFalke Thanks for reviewing. Addressed your optimization request and dropped all changes that are unrelated to `CAddrMan`. Should hopefully be ready for merge now :)

MarcoFalke commented on src/test/fuzz/addrman.cpp:
```diff
@@ -0,0 +1,121 @@
+    if (fuzzed_data_provider.ConsumeBool()) {
+        while (fuzzed_data_provider.ConsumeBool()) {
+            addr_man.m_asmap = BytesToBits(ConsumeRandomLengthByteVector(fuzzed_data_provider));
+        }
```
any reason to re-assign this in a loop?

MarcoFalke replied:
Also to make seeds re-usable it should probably use the same encoding like the other places. `git grep 'asmap = ConsumeRan'`

practicalswift replied:
Oh, the loop was a bug. Sorry about that! Thanks for catching.

Now using the newly introduced `ConsumeRandomLengthBitVector(...)` also in `src/test/fuzz/net`.

practicalswift committed:
fuzz: Use ConsumeRandomLengthBitVector(...) in src/test/fuzz/connman and src/test/fuzz/net

practicalswift commented:
@MarcoFalke Thanks for re-reviewing! Addressed all feedback. Should hopefully be ready for final review _now_ :)

MarcoFalke commented on src/test/fuzz/net.cpp:
```diff
@@ -63,7 +63,7 @@ void test_one_input(const std::vector<uint8_t>& buffer)
         }
         case 3: {
-            const std::vector<bool> asmap = ConsumeRandomLengthIntegralVector<bool>(fuzzed_data_provider, 128);
+            const std::vector<bool> asmap = ConsumeRandomLengthBitVector(fuzzed_data_provider);
```
please also adjust connman with this patch

practicalswift commented:
@MarcoFalke Thanks for re-re-reviewing! Addressed all feedback. Should hopefully be ready for final final review now :)

MarcoFalke commented:
review ACK d04a17a7907c57f7b570e1b9743fd63489bdad68
